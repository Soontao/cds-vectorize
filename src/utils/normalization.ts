import { Vector } from "../encoders/type.js";
import { fillMissingValues, isNull } from "../encoders/util.js";

/**
 * min-max value normalization
 *
 * @param vectors
 * @returns
 */
export function normalization(vectors: Array<Vector>): Array<Vector> {
  if (vectors.length === 0 || isNull(vectors)) {
    return [];
  }
  vectors = vectors.map(fillMissingValues);
  const numFeatures = vectors[0]?.length;

  // Calculate min and max values for each feature
  const minValues = new Array(numFeatures);
  const maxValues = new Array(numFeatures);

  for (const vector of vectors) {
    vector.forEach((value, index) => {
      if (minValues[index] == undefined) {
        minValues[index] = value;
      } else {
        minValues[index] = Math.min(minValues[index], value);
      }
      if (maxValues[index] == undefined) {
        maxValues[index] = value;
      } else {
        maxValues[index] = Math.max(maxValues[index], value);
      }
    });
  }

  // Apply Min-Max Scaling to each vector
  const normalizedVectors = vectors.map((vector) => {
    const newVec = vector.map((value, index) => {
      if (minValues[index] == maxValues[index]) {
        return minValues[index];
      }
      return (value - minValues[index]) / (maxValues[index] - minValues[index]);
    });
    Object.defineProperty(newVec, "__original", { value: vector, enumerable: false });
    return newVec;
  });

  return normalizedVectors;
}

export default normalization;
