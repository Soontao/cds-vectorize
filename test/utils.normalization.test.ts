// @ts-nocheck
import { normalization } from "../src/utils/normalization.js";

describe("Normalization Test Suite", () => {
  it("should normalize vectors using Min-Max Scaling", () => {
    const vectors = [
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ];

    const normalizedVectors = normalization(vectors);

    // Ensure the normalized vectors have values between 0 and 1
    normalizedVectors.forEach((vector) => {
      vector.forEach((value) => {
        expect(value).toBeGreaterThanOrEqual(0);
        expect(value).toBeLessThanOrEqual(1);
      });
    });
    expect(normalizedVectors[0].__original).toEqual(vectors[0]);
    expect(normalizedVectors).toMatchSnapshot();
  });

  it("should handle empty vectors array", () => {
    const vectors: number[][] = [];
    const normalizedVectors = normalization(vectors);

    // Expect the result to be an empty array
    expect(normalizedVectors).toEqual([]);
  });

  it("should handle vectors with null values", () => {
    const vectors = [
      [1, 2, 3],
      [null, 5, 6],
      [7, 8, null],
    ];

    // Normalize vectors with null values
    const normalizedVectors = normalization(vectors);

    expect(normalizedVectors).toMatchSnapshot();
  });

  it("should handle vectors with random values", () => {
    const vectors = [
      [Math.random(), Math.random(), Math.random()],
      [Math.random(), Math.random(), Math.random()],
      [Math.random(), Math.random(), Math.random()],
    ];

    // Normalize vectors with random values
    const normalizedVectors = normalization(vectors);

    // Expect the result to have values between 0 and 1
    normalizedVectors.forEach((vector) => {
      vector.forEach((value) => {
        expect(value).toBeGreaterThanOrEqual(0);
        expect(value).toBeLessThanOrEqual(1);
      });
    });
  });

  it("should handle vectors with all zero values", () => {
    const vectors = [
      [0, 0, 0],
      [0, 0, 0],
      [0, 0, 0],
    ];

    // Normalize vectors with all zero values
    const normalizedVectors = normalization(vectors);

    // Expect the result to have values of 0 for all elements
    normalizedVectors.forEach((vector) => {
      vector.forEach((value) => {
        expect(value).toEqual(0);
      });
    });
    expect(normalizedVectors).toMatchSnapshot();
  });
});
