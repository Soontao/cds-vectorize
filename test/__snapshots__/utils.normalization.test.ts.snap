// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Normalization Test Suite should handle vectors with all zero values 1`] = `
[
  [
    0,
    0,
    0,
  ],
  [
    0,
    0,
    0,
  ],
  [
    0,
    0,
    0,
  ],
]
`;

exports[`Normalization Test Suite should handle vectors with null values 1`] = `
[
  [
    0.9999084570434676,
    0,
    0.9999542278233804,
  ],
  [
    0,
    0.5,
    1,
  ],
  [
    1,
    1,
    0,
  ],
]
`;

exports[`Normalization Test Suite should normalize vectors using Min-Max Scaling 1`] = `
[
  [
    0,
    0,
    0,
  ],
  [
    0.5,
    0.5,
    0.5,
  ],
  [
    1,
    1,
    1,
  ],
]
`;
